src/sql/workbench/services/notebook/browser/notebookServiceImpl.ts

```diff
@@ -248,30 +248,60 @@ export class NotebookService extends Disposable implements INotebookService {
 		lifecycleService.onWillShutdown(() => this.shutdown());
 	}
 
-	public async createNotebookInput(options: INotebookShowOptions, resource?: UriComponents): Promise<IEditorInput | undefined> {
+	private getUntitledFileUri(): URI {
+		// Need to create a new untitled URI, so find the lowest numbered one that's available
+		let uri: URI;
+		let counter = 1;
+		do {
+			uri = URI.from({ scheme: Schemas.untitled, path: `Notebook-${counter}` });
+			counter++;
+		} while (this._untitledEditorService.get(uri));
+		return uri;
+	}
+
+	public async createNotebookInputFromContents(providerId: string, contents?: nb.INotebookContents, resource?: UriComponents): Promise<IEditorInput> {
 		let uri: URI;
 		if (resource) {
 			uri = URI.revive(resource);
 		} else {
-			// Need to create a new untitled URI, so find the lowest numbered one that's available
-			let counter = 1;
-			do {
-				uri = URI.from({ scheme: Schemas.untitled, path: `Notebook-${counter}` });
-				counter++;
-			} while (this._untitledEditorService.get(uri));
+			uri = this.getUntitledFileUri();
+			resource = uri;
+		}
+
+		let serializedContent: string;
+		if (contents) {
+			// Have to serialize contents again first, since our notebook code assumes input is based on the raw file contents
```

Comment:
Before, this method was being passed the stringified version of a NotebookContents object, which is the wrong way to do it. Our code expects the initialContent in NotebookShowOptions to be the actual raw file contents, but in the case of .dib files the file contents aren't stored as JSON. So we have to serialize the NotebookContents back into the original raw file format before passing it to createNotebookInput. We're forced to do it this way because the VS Code API for creating a notebook is passed a generic notebook JSON object (vscode.NotebookData), so we have to rely on a bit of conversion in the middle.